### ISSUE TITLE
Add single-shot measurements

### ISSUE THREAD

**Issue by co9olguy** [XanaduAI/pennylane#171]
Machine learning libraries allow for ops that return samples from a distribution (rather than averages/expectations). This is also a common occurance in quantum algorithms. This might be a helpful feature to add to PennyLane to expand the scope of what it can do. 

Of course, these operations will not be natively differentiable (which is also the case in e.g., TensorFlow), so we will need to throw a warning when any kind of gradient is requested for such ops.

**Comment by co9olguy** [comment]
A use-case to keep in mind is where a circuit is trained in a differentiable way (using expvals), but later we want to sample the fullly trained circuit in a single-shot fashion

**Comment by josh146** [comment]
My thinking on this:

1. We actually already support single shot measurements at the device level, for instance by specifying the keyword argument ``shots=1``. So what you could do is create two separate 'devices': one for expectation values/training, and one for single shot samples, and use them where required:
    ```python
    dev_train = qml.device('forest.qpu', device='Aspen-1-2Q-A', shots=1024)
    dev_sample = qml.device('forest.qpu', device='Aspen-1-2Q-A', shots=1)

    def circuit(..):
        ...
        return qml.expval.PauliZ(0), qml.expval.PauliZ(1)
    
    qnode = qml.QNode(circuit, dev_train)
    # training steps here...
    
    # after training, generate samples
    qnode_sample = qml.QNode(circuit, dev_sample)
    qnode_sample(final_weights)
    ```
    Note that `qnode_sample` is still 'differentiable' by the rules of PennyLane, but the gradients will be significantly error-prone due to the low number of shots.

2. An alternate proposal: introduce a new class of operations `qml.sample`, similar to `qml.expval` and `qml.ops`:
    ```python
    dev_train = qml.device('forest.qpu', device='Aspen-1-2Q-A', shots=1024)

    @qml.qnode(dev)
    def circuit(..):
        ...
        return qml.sample.PauliZ(0), qml.sample.PauliZ(1)
    ```

    This has two advantages:

    a. the return value for each sample command will be a _list_ of samples, of size `circuit_out*dev_shots`:
    ```python
    >>> circuit(0.43)
    [[0, 0, 1, 0, 1, ...., 1], [0, 1, 0, 0, 1, 1, ..., 0]]
    ```

    b. attempting to differentiate such a qnode will result in an exception raised.

Thoughts, @mariaschuld, @co9olguy?

Option (1) already exists today, and the documentation will just have to be updated so that people know how to use it. Option (2) would require the `qml.sample` submodule to be added to PennyLane, the plugin API be updated, and all plugins updated.

**Comment by co9olguy** [comment]
Note: this will be implemented as a follow-up to the `expval` refactor #232

**Comment by josh146** [comment]
This is done now that #256  is merged.
